### ISSUE TITLE
Review new behavior of the LintDiff PatchBodyParameterSchema rule

### ISSUE THREAD

**Issue by konrad-jamrozik** [Azure/azure-sdk-tools#7771]
In this PR @mikekistler fixed a bug with the rule:
- https://github.com/Azure/azure-openapi-validator/pull/661

This led to the rule behavior change, making it more strict.

We need to review its new behavior.

Relevant [Teams discussion](https://teams.microsoft.com/l/message/19:4f661242c446452e895359cc3ef45125@thread.tacv2/1708933290767?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=0cab4ce9-7691-42ae-82e3-460d4346a710&parentMessageId=1708933290767&teamName=ARM%20API%20Reviewers&channelName=ARM%20linter%20rules&createdTime=1708933290767)

Relevant comment:
- https://github.com/Azure/azure-rest-api-specs/pull/27888#issuecomment-1963573749

**Comment by konrad-jamrozik** [comment]
@mikekistler could you provide more context on the new rule behavior?

**Comment by mikekistler** [comment]
Previously the rule checked a PATCH request body for any properties that are 1) required, 2) have a default value, or 2) have `x-ms-mutability` of ["create"].

My change simply enforces these same rules on nested properties as well.

The updated rule flagged a PATCH operation that, two-levels deep in the body, has a "sku" property whose schema is an object with a single property, "name", that is required.

I think that flagging a required property _anywhere_ in a PATCH request body is consistent with the spirit of this rule.

**Comment by rkmanda** [comment]
yup, agree that Patch request body must not have a required property at any level

**Comment by rkmanda** [comment]
I actually have a question that popped up regarding this. When a property is marked required at a particular nesting level in swagger, does it actually make it required for the entire payload or does it make it required only when the parent property is present on the wire? I will check what the open API spec language says about this and get back. @konrad-jamrozik , @mikekistler

**Comment by rkmanda** [comment]
![image](https://github.com/Azure/azure-sdk-tools/assets/16844618/283db194-f755-4f15-b4c3-638d94bb5529)

**Comment by rkmanda** [comment]
However, from a patch standpoint, the argument is that marking a property "required" at a particular level, makes it mandatory for that property to always be specified when a Patch is being used to update its siblings. So I guess we are still good with the change we made. One issue I see with this is a case where a set of properties are always required together at the same level, then theres no way to express that for Patch APIs.

**Comment by mikekistler** [comment]
@rkmanda I had to think through this exact scenario too. I think the general principle is that you can't use validation on the request payload _alone_ to ensure that the request is valid -- it will depend on the state of the service as well. These are just special cases of this.

**Comment by konrad-jamrozik** [comment]
This rule is causing some trouble. See email thread with subject: `RE: [LintDiff] common-types causes PatchBodyParametersSchema`.

**Comment by mikeharder** [comment]
Spec authors continue to be confused by failures of this rule, and are starting to call the errors "false positives": 
 https://github.com/Azure/azure-sdk-tools/issues/7802

@markcowl: In https://github.com/Azure/typespec-azure/issues/383, you seem to suggest this is a bug in typespec-autorest, rather than azure-openapi-validator.  If so, I think this issue and #7802 should both be closed, and we track this only as a bug in TypeSpec?  Do you agree?

**Comment by mikekistler** [comment]
+1 to closing this issue and #7802.

Fixing the TypeSpec issue will help but I expect some teams will still get tripped up on this.

**Comment by konrad-jamrozik** [comment]
@mikekistler so if I understand correctly your recommendation is:

- Close #7771 and #7802 as "do nothing more"
- The new strict behavior as given by https://github.com/Azure/azure-openapi-validator/pull/661 is the correct one

?

**Comment by mikeharder** [comment]
From this discussion with @markcowl:

https://teams.microsoft.com/l/message/19:4f661242c446452e895359cc3ef45125@thread.tacv2/1718209098071?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=0cab4ce9-7691-42ae-82e3-460d4346a710&parentMessageId=1718209098071&teamName=ARM%20API%20Reviewers&channelName=ARM%20linter%20rules&createdTime=1718209098071

I believe the root cause is between azure-openapi-validator and the swagger in common-types.  According to Mark, this issue can impact specs using either TypeSpec or handwritten-swagger, so I think we can simplify by removing TypeSpec from the repro.

I believe the root cause can be summarized as "Rule PatchBodyParameterSchema is incompatible with the shared swagger in common-types"?  So either the rule or common-types needs to change.  We think the correct fix is to change common-types, but as a temporary workaround, reverting the rule change will be easier.

**Comment by mikekistler** [comment]
I'm skeptical that all the instances of this rule firing are caused by the common-types.

And when they are caused by the common types, can't we simply use suppressions until the common types are fixed?
